discovery/consul/consul.go

```diff
@@ -226,94 +293,142 @@ func (d *Discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
-			if !d.shouldWatch(name) {
-				continue
-			}
-			if _, ok := services[name]; ok {
-				continue // We are already watching the service.
-			}
-
-			srv := &consulService{
-				client: d.client,
-				name:   name,
-				labels: model.LabelSet{
-					serviceLabel:    model.LabelValue(name),
-					datacenterLabel: model.LabelValue(d.clientDatacenter),
-				},
-				tagSeparator: d.tagSeparator,
-				logger:       d.logger,
-			}
-
-			wctx, cancel := context.WithCancel(ctx)
-			go srv.watch(wctx, ch)
+// Watch the catalog for new services we would like to watch. This is called only
+// when we don't know yet the names of the services and need to ask Consul the
+// entire list of services.
+func (d *Discovery) watchServices(ctx context.Context, ch chan<- []*targetgroup.Group, lastIndex *uint64, services map[string]func()) error {
+	catalog := d.client.Catalog()
+	level.Debug(d.logger).Log("msg", "Watching services", "tag", d.watchedTag)
+
+	t0 := time.Now()
+	srvs, meta, err := catalog.Services(&consul.QueryOptions{
+		WaitIndex:  *lastIndex,
+		WaitTime:   watchTimeout,
+		AllowStale: d.allowStale,
+		NodeMeta:   d.watchedNodeMeta,
+	})
+	elapsed := time.Since(t0)
+	rpcDuration.WithLabelValues("catalog", "services").Observe(elapsed.Seconds())
 
-			services[name] = cancel
+	if err != nil {
+		level.Error(d.logger).Log("msg", "Error refreshing service list", "err", err)
+		rpcFailuresCount.Inc()
+		time.Sleep(retryInterval)
+		return err
+	}
+	// If the index equals the previous one, the watch timed out with no update.
+	if meta.LastIndex == *lastIndex {
+		return nil
+	}
+	*lastIndex = meta.LastIndex
+
+	// Check for new services.
+	for name := range srvs {
+		// catalog.Service() returns a map of service name to tags, we can use that to watch
+		// only the services that have the tag we are looking for (if specified).
+		// In the future consul will also support server side for service metadata.
+		// https://github.com/hashicorp/consul/issues/1107
+		if !d.shouldWatch(name, srvs[name]) {
+			continue
+		}
+		if _, ok := services[name]; ok {
+			continue // We are already watching the service.
 		}
 
-		// Check for removed services.
-		for name, cancel := range services {
-			if _, ok := srvs[name]; !ok {
-				// Call the watch cancelation function.
-				cancel()
-				delete(services, name)
-
-				// Send clearing target group.
-				select {
-				case <-ctx.Done():
-					return
-				case ch <- []*targetgroup.Group{{Source: name}}:
-				}
+		wctx, cancel := context.WithCancel(ctx)
+		d.watchService(wctx, name, ch)
+		services[name] = cancel
+	}
+
+	// Check for removed services.
+	for name, cancel := range services {
+		if _, ok := srvs[name]; !ok {
+			// Call the watch cancellation function.
+			cancel()
+			delete(services, name)
+
+			// Send clearing target group.
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case ch <- []*targetgroup.Group{{Source: name}}:
 			}
 		}
 	}
+	if elapsed < d.refreshInterval {
```

Comment:
This would be easier with a ticker up in Run

Reply:
Yes. But I'm not sure how to do the same thing for `watch()` which does the same thing (since a ticker is for a single "reader"). Maybe https://github.com/VividCortex/multitick but not sure we want a dependency on that.

Reply:
You'd have a ticker per watch()